### ISSUE TITLE
Support clang's new shell autocompletion feature

### ISSUE THREAD

**Issue by Teemperor** [fish-shell/fish-shell#4174]
The next version of clang will ship with a `--autocomplete` flag that provides a backend for shells to complete all flags that the current clang binary supports. The general idea is that the shell passes the flag it currently wants to complete to clang via`clang --autocomplete=-fsyntax` and then clang prints a list of flags that this completes to:

```js
$./bin/clang --autocomplete=-std
-std-default= -std: -std= -stdlib=
```

This also works with all the values that the flags take, e.g. in this case return all the stdlibs clang supports in its `-stdlib=libc++` flag:

```js
$ ./bin/clang --autocomplete=-stdlib=,
libc++ libstdc++ platform
```

For a better overview, see the clang's bash completion script that illustrates the way the `--autocomplete` flag works: [bash-completions.sh](https://github.com/llvm-mirror/clang/blob/master/utils/bash-autocomplete.sh).

The big advantage is that this allows to have a single shell completion that works with all flags of all future clang versions (even the custom vendor versions from Apple or a self-build clang from trunk!), as the flag just queries the internal option parsing logic of the selected clang binary. 

It would be very useful if fish would use this feature and also start calling this flag when trying to complete a clang flag.

**Comment by faho** [comment]
I'd be willing to do this once I have that version on my machine. Any idea when it's going to be released?

Some notes for later:

- It looks like this doesn't use the bash completion variables ($COMP_LINE and such), which is nice. If it does, there's already an implementation in the npm completions.

- This seems to expect splitting on space - so it needs a `string split " "`.

- I'm not sure about the `--autocomplete=-stdlib=,` one - it should ultimately result in `-stdlib=libc++` being offered. It might need manual translation.

**Comment by Teemperor** [comment]
LLVM/clang 5 will release on 23rd of August, but it's really up to your OS-vendor when it arrives on your machine :). If there are some API changes with the --autocomplete flag that can't be worked around from the fish complete script, then we can still influence the API for the script until the 19th of July when clang 5 is tagged (but no promises on this).

**Comment by faho** [comment]
For fish, this would ideally receive the full clang commandline up to the cursor. I.e. we'd just call `clang --autocomplete=(commandline -pc)` ("-p" only selects the current "process" without any other parts in a pipe and such, "-c" only prints up to the cursor). That way it would also have the full context - which other options have already been specified and such.

Also, it would delimit different candidates with newlines, not spaces - this shouldn't break anything for bash either.

Also, it would use `--autocomplete=-stdlib=` to suggest options for "-stdlib=" (without the trailing ","), and it would print the entire thing - "-stdlib=libc++", not just the "libc++" part.

Support for descriptions would be nice to have - something like `-stdlib=\tSelect a standard library`, but that would interact badly with other shells.

None of these are dealbreakers, though.

**Comment by zanchey** [comment]
Perhaps the clang maintainers would consider shipping the fish completion themselves - it might be worth trying to get it written before the release.

**Comment by Teemperor** [comment]
The autocomplete support has landed in clang 5.0 which will release in two weeks and is currently in testing and available in the [LLVM apt repository](http://apt.llvm.org/). We added support for flag descriptions in the format fish wants (separated by `\t`) and moved to newlines as separators. The `,`'s between the values however stayed, sorry :).

The [clang unit tests](https://github.com/llvm-mirror/clang/blob/master/test/Driver/autocomplete.c) should help with explaining the API. Also ping for @yamaguchi1024 who is the code owner of this cool feature, she maybe has time to answer your questions.

@zanchey Shipping this is up to the distribution maintainers, which is quite a large and diverse group of people we would have to get on board for this to work out :/

**Comment by faho** [comment]
>We added support for flag descriptions in the format fish wants (separated by \t) and moved to newlines as separators. 

Nice!

>The ,'s between the values however stayed, sorry :).

There's probably a reason for them that I don't get. Anyway, we can deal with that.
